jolt-evm-verifier/src/reference/JoltVerifier.sol

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.8.21;
+
+import {IVerifier} from "../interfaces/IVerifier.sol";
+import {Transcript, FiatShamirTranscript} from "../subprotocols/FiatShamirTranscript.sol";
+import {Fr, FrLib, sub, MODULUS} from "./Fr.sol";
+import {Jolt} from "./JoltTypes.sol";
+import {UniPoly, UniPolyLib} from "./UniPoly.sol";
+
+import "forge-std/console.sol";
+
+error GrandProductArgumentFailed();
+error SumcheckFailed();
+
+contract JoltVerifier is IVerifier {
+    using FiatShamirTranscript for Transcript;
+
+    function verifySumcheckLayer(
+        Jolt.BatchedGrandProductLayerProof memory layer,
+        Transcript memory transcript,
+        Fr claim,
+        uint256 degree_bound,
+        uint256 num_rounds
+    ) internal pure returns (Fr, Fr[] memory) {
+        if (layer.sumcheck_univariate_coeffs.length != num_rounds) {
+            revert SumcheckFailed();
+        }
+
+        Fr e = claim;
+        Fr[] memory r = new Fr[](num_rounds);
+
+        for (uint256 i = 0; i < num_rounds; i++) {
+            UniPoly memory poly = UniPolyLib.decompress(layer.sumcheck_univariate_coeffs[i], e);
+
+            //verify degree bound
+            if (layer.sumcheck_univariate_coeffs[i].length != degree_bound) {
+                revert SumcheckFailed();
+            }
+
+            // check if G_k(0) + G_k(1) = e
+            if (UniPolyLib.evalAtZero(poly) + UniPolyLib.evalAtOne(poly) != e) {
+                revert SumcheckFailed();
+            }
+
+            // TODO - We can move this into the transcript lib
+            transcript.append_bytes32("UniPoly_begin");
+            for (uint256 j = 0; j < poly.coeffs.length; j++) {
+                transcript.append_scalar(Fr.unwrap(poly.coeffs[j]));
+            }
+            transcript.append_bytes32("UniPoly_end");
+
+            // Sample random from the transcript
+            Fr r_i = Fr.wrap(transcript.challenge_scalar(MODULUS));
+            r[i] = r_i;
+
+            e = UniPolyLib.evaluate(poly, r_i);
+        }
+        return (e, r);
+    }
+
+    function buildEqEval(Fr[] memory rGrandProduct, Fr[] memory rSumcheck) internal pure returns (Fr eqEval) {
```

Comment:
`evalEqMLE` might be a more apt name.

`buildEqEval` implies it may be building the `Eq` evaluation table.